lib/babe/verify.go

```diff
@@ -18,13 +18,210 @@ package babe
+func (v *VerificationManager) VerifyBlock(header *types.Header) (bool, error) {
+	fromEpoch, err := v.isBlockFromEpoch(header.Hash(), v.currentEpoch)
+	if err != nil {
+		return false, err
+	}
+
+	digest, err := checkForConsensusDigest(header)
+	if err != nil {
+		return false, err
+	}
+
+	ok, err := v.verifier.verifyAuthorshipRight(header)
+	if err != nil {
+		return false, err
+	}
+
+	if digest == nil {
+		// verify and return
+		return ok, nil
+	}
+
+	// check if first block has been set for current epoch
+	if fromEpoch && v.firstBlock != nil {
+
+		// check if block header has lower block number than current first block
+		if header.Number.Cmp(v.firstBlock.Number) < 0 {
+			v.firstBlock = header
+		}
+
+	} else if fromEpoch {
+		// set first block in current epoch
+		v.firstBlock = header
+	}
+
+	return ok, nil
+}
+
+// blockFromCurrentEpoch verifies the provided block hash is from given epoch
+func (v *VerificationManager) isBlockFromEpoch(hash common.Hash, epoch uint64) (bool, error) {
+	// get epoch number of block header
+	blockEpoch, err := v.getBlockEpoch(hash)
+	if err != nil {
+		return false, fmt.Errorf("[babe verifier] failed to get epoch from block header: %s", err)
+	}
+
+	// check if block epoch number matches given epoch number
+	if blockEpoch != epoch {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// getBlockEpoch gets the epoch number using the provided block hash
+func (v *VerificationManager) getBlockEpoch(hash common.Hash) (epoch uint64, err error) {
+	// get slot number to determine epoch number
+	// TODO: directly calulcate this from the pre-digest
+	slot, err := v.blockState.GetSlotForBlock(hash)
+	if err != nil {
+		return epoch, fmt.Errorf("failed to get slot from block hash: %s", err)
+	}
+
+	if slot != 0 {
+		// epoch number = (slot - genesis slot) / epoch length
+		epoch = (slot - 1) / 6 // TODO: use epoch length from babe or core config
+	}
+
+	return epoch, nil
+}
+
+// checkForConsensusDigest returns a consensus digest from the header, if it exists.
+func checkForConsensusDigest(header *types.Header) (*types.ConsensusDigest, error) {
+	// check if block header digest items exist
+	if header.Digest == nil || len(header.Digest) == 0 {
+		return nil, fmt.Errorf("header digest is not set")
+	}
+
+	// declare digest item
+	var consensusDigest *types.ConsensusDigest
+
+	// decode each digest item and check its type
+	for _, digest := range header.Digest {
+		item, err := types.DecodeDigestItem(digest)
+		if err != nil {
+			return nil, err
+		}
+
+		// check if digest item is consensus digest type
+		if item.Type() == types.ConsensusDigestType {
+			var ok bool
+			consensusDigest, ok = item.(*types.ConsensusDigest)
+			if ok {
+				break
+			}
+		}
+	}
+
+	return consensusDigest, nil
+}
+
+// Verifier represents a BABE verifier for a specific epoch
+type Verifier struct {
+	blockState    BlockState
+	authorityData []*AuthorityData
+	randomness    byte // TODO: update to [32]byte when runtime is updated
```

Comment:
ok, I see. We should try to reuse comments as much as possible and maybe add issue numbers. I see you reuse this one here in this file and the same could be used here ->

https://github.com/ChainSafe/gossamer/pull/778/files#diff-4fc1ad0d25a69aae6934773bd9823be2R215